src/Language/PureScript/TypeChecker/Entailment.hs

```diff
@@ -80,139 +120,278 @@ entails shouldGeneralize moduleName context = solve
+                modify (combineContexts newContext)
+                -- Mark this constraint for generalization
+                tell (mempty, [(ident, unsolved)])
+                return (Var qident)
+              Deferred ->
+                -- Constraint was deferred, just return the dictionary unchanged,
+                -- with no unsolved constraints. Hopefully, we can solve this later.
+                return (TypeClassDictionary (Constraint className' tys'' conInfo) context hints)
+          where
+            -- | When checking functional dependencies, we need to use unification to make
+            -- sure it is safe to use the selected instance. We will unify the solved type with
+            -- the type in the instance head under the substition inferred from its instantiation.
+            -- As an example, when solving MonadState t0 (State Int), we choose the
+            -- MonadState s (State s) instance, and we unify t0 with Int, since the functional
+            -- dependency from MonadState dictates that t0 should unify with s\[s -> Int], which is
+            -- Int. This is fine, but in some cases, the substitution does not remove all TypeVars
+            -- from the type, so we end up with a unification error. So, any type arguments which
+            -- appear in the instance head, but not in the substitution need to be replaced with
+            -- fresh type variables. This function extends a substitution with fresh type variables
+            -- as necessary, based on the types in the instance head.
+            withFreshTypes
+              :: TypeClassDictionaryInScope
+              -> Matching Type
+              -> m (Matching Type)
+            withFreshTypes TypeClassDictionaryInScope{..} subst = do
+                let onType = everythingOnTypes S.union fromTypeVar
+                    typeVarsInHead = foldMap onType tcdInstanceTypes
+                                  <> foldMap (foldMap (foldMap onType . constraintArgs)) tcdDependencies
+                    typeVarsInSubst = S.fromList (M.keys subst)
+                    uninstantiatedTypeVars = typeVarsInHead S.\\ typeVarsInSubst
+                newSubst <- traverse withFreshType (S.toList uninstantiatedTypeVars)
+                return (subst <> M.fromList newSubst)
+              where
+                fromTypeVar (TypeVar v) = S.singleton v
+                fromTypeVar _ = S.empty
+
+                withFreshType s = do
+                  t <- freshType
+                  return (s, t)
+
+            unique :: [(a, TypeClassDictionaryInScope)] -> m (EntailsResult a)
+            unique [] | deferErrors = return Deferred
+                      -- We need a special case for nullary type classes, since we want
+                      -- to generalize over Partial constraints.
+                      | shouldGeneralize && (null tys' || any canBeGeneralized tys') = return (Unsolved con')
```

Comment:
@garyb Here is the generalization check. Note the change to use `any` instead of `all`.